rust/src/push/base_rules.rs

```diff
@@ -596,6 +610,68 @@ pub const BASE_APPEND_UNDERRIDE_RULES: &[PushRule] = &[
         default: true,
         default_enabled: true,
     },
+    PushRule {
```

Comment:
Does the placement in the order here matter? The MSC doesn't state where it goes.

Reply:
Hmm, the ordering of these rules will matter yes. We want the `*one_to_one` rules to be tested in clients first before their group room counterparts. I'm a little confused on how priority works in push rules though. The spec [seems to imply](https://spec.matrix.org/v1.5/client-server-api/#push-rules) that rules are checked by kind, then an "ordering priority". I'm guessing the latter is the order that the rules are sent to clients in, rather than a field? It's a bit vague though.

We can also just sidestep the ordering requirement by placing a `room_member_count>2` condition on the group room rules, but that shouldn't be necessary.

Reply:
Note that I have asserted the ordering in the Complement tests: https://github.com/matrix-org/complement/pull/578/files#diff-ce4102a82b595c7b72d6f945db418fbc7d34ff7973f5ca6e8935ad98a09273d2R161

Reply:
> Hmm, the ordering of these rules will matter yes.

Sorry for being unclear! I meant how did you decide this went after `global/underride/.im.vector.jitsi`.

I think internally the ordering is reasonable and matches the MSC.

Reply:
Oh! I didn't consider ordering against other `underride` push rules. But perhaps this should go above `global/underride/.m.rule.encrypted`, such that clients match this push rule before the generic "encrypted message" one?

Though that does make me question all of the other non-state event-related rules in this file. Shouldn't they go above rules that match generic `m.room.encrypted` events? I suppose the client must be doing some mangling of the order here/filtering out of the `m.rule.encrypted` rule(s) already?

It turns out matrix-react-sdk is just [sorting known rules](https://github.com/matrix-org/matrix-react-sdk/blob/cce506b1c142bebdaf4c548fb50e11fa3e00b861/src/components/views/settings/Notifications.tsx#L288-L298) client-side anyways into [this order](https://github.com/matrix-org/matrix-react-sdk/blob/1daa34b51c6c25724c4bceb3f630d2090557bf83/src/components/views/settings/Notifications.tsx#L70-L88). Unknown rules appear at the end in the order they were received from the homeserver. Interestingly the generic `EncryptedMessage` and `EncryptedDM` rules are put in front. I'm not sure how this achieves the desired result unless there's additional logic I'm missing.

Reply:
My understanding was the spec mandates an order (and thus MSCs must define an order).

Reply:
Have tried to get some clarity in https://github.com/matrix-org/matrix-spec-proposals/pull/3930#discussion_r1052364430.

Reply:
We've been given the go ahead to define our own ordering for now: https://github.com/matrix-org/matrix-spec-proposals/pull/3930#discussion_r1073779883